azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,146 @@
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = Queue()
+
+    @staticmethod
+    def get_thread_name():
+        return SendTelemetryEventsHandler._THREAD_NAME
+
+    def run(self):
+        logger.info("Start Extension Telemetry service.")
+        self.start()
+
+    def is_alive(self):
+        return self._thread is not None and self._thread.is_alive()
+
+    def start(self):
+        self._thread = threading.Thread(target=self._process_telemetry_thread)
+        self._thread.setDaemon(True)
+        self._thread.setName(self.get_thread_name())
+        self._thread.start()
+
+    def stop(self):
+        """
+        Stop server communication and join the thread to main thread.
+        """
+        self.should_run = False
+        # Set the event to unblock the thread to ensure that the thread is not blocking shutdown.
+        self._should_process_events.set()
+        if self.is_alive():
+            self.join()
+
+    def join(self):
+        self._queue.join()
+        self._thread.join()
+
+    def stopped(self):
+        return not self.should_run
+
+    def enqueue_event(self, event):
+        # Add event to queue and set event
+        if self.stopped():
+            raise ServiceStoppedError("{0} is stopped, not accepting anymore events".format(self.get_thread_name()))
+
+        # Queue.put() can block if the queue is full which can be an uninterruptible wait. Blocking for a max of
+        # SendTelemetryEventsHandler._MAX_TIMEOUT seconds and raising a ServiceStoppedError to retry later.
+        try:
+            self._queue.put(event, timeout=SendTelemetryEventsHandler._MAX_TIMEOUT)
+        except Full as error:
+            raise ServiceStoppedError(
+                "Queue full, stopping any more enqueuing until the next run. {0}".format(ustr(error)))
+
+        # Set the event if any enqueue happens (even if already set) to trigger sending those events
+        self._should_process_events.set()
+
+    def _process_telemetry_thread(self):
+        logger.info("Successfully started the {0} thread".format(self.get_thread_name()))
+        try:
+            # On demand wait, start processing as soon as there is any data available in the queue. In worst case,
+            # also keep checking every SendTelemetryEventsHandler._MAX_TIMEOUT secs to avoid uninterruptible waits
+            while not self.stopped():
+                self._should_process_events.wait(timeout=SendTelemetryEventsHandler._MAX_TIMEOUT)
+                self._send_events_in_queue()
+
+        except Exception as error:
+            err_msg = "An unknown error occurred in the {0} thread main loop, stopping thread. Error: {1}, Stack: {2}".format(
+                self.get_thread_name(), ustr(error), traceback.format_exc())
+            add_event(op=WALAEventOperation.UnhandledError, message=err_msg, is_success=False)
+
+    def _get_events_in_queue(self):
+        while not self._queue.empty():
```

Comment:
should this be checking for stopped() as well?

Reply:
I specifically didnt add it here. The whole idea is that we should clear the queue even if the service is asked to stop to avoid data loss. We won't let anyone enqueue anymore events if `self.stopped` is True ensuring that the queue is not never ending.